### PR TITLE
Add filtering framework and exponential filter

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Adds a function compute an exponential filter.
- Adds an action to apply the exponential filter to specific tags.

This was needed to get spherical Bondi accretion stable using a spherical shell domain. Here is a plot of the errors to show this stabilizes the scheme:

plot

Depends loosely on:
- [x] #1489 

The commits that are *NOT* to be reviewed as part of this PR are:
- Remove break statements that don't do anything
- Fix TensorExpressions linking errors with lld 8
- Rename grid_points_to_spectral_matrix and inverse functions

The order of commits that *are* part of this PR:
1. Add function to compute exponential filter matrix
2. Add exponential filter action

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+ * The exponential filter rescales the modal coefficients \f$c_i\f$ as:
+ *
+ * \f{align*}{
+ *  c_i\to c_i \exp\left[-\alpha \left(\frac{i}{N}\right)^{2m}\right]
+ * \f}
+ *
+ * where \f$N\f$ is the basis order, \f$\alpha\f$ determines much the
+ * coefficients are rescaled, and \f$m\f$ (given by the `HalfPower` option)
+ * determines how aggressive/broad the filter is (lower values means filtering
+ * more coefficients). Setting \f$\alpha=36\f$ results in setting the highest
```
Comment (moxcodes) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: 'setting' here might be confusing to a newcomer. Perhaps 'zeroing' or 'setting to zero' or 'setting to machine epsilon (near zero)'

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) noexcept;
+
+  bool operator==(const ExponentialFilter& rhs) const noexcept;
+
+ private:
+  template <size_t Dim>
+  const Matrix& filter_matrix(const Mesh<Dim>& mesh, size_t d) const noexcept {
+    switch (mesh.basis(d)) {
```
Comment (moxcodes) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: Can these not be extracted from mesh type aliases? On a cursory inspection, it seems there aren't any such type aliases, but it's not clear to me why we shouldn't have them.

File: src/NumericalAlgorithms/Spectral/Filtering.hpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+/// \endcond
+
+namespace Spectral {
+/*!
+ * \ingroup SpectralGroup
+ * \brief Matrices for filtering spectral coefficients.
+ */
+namespace filtering {
+/*!
+ * \brief Returns a `Matrix` by which to multiply the modal coefficients to
```
Comment (moxcodes) [inline src/NumericalAlgorithms/Spectral/Filtering.hpp]: modal or nodal?

File: src/NumericalAlgorithms/Spectral/Filtering.hpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * The exponential filter rescales the coefficients according to:
+ *
+ * \f{align*}{
+ *  c_i\to c_i \exp\left[-\alpha \left(\frac{i}{N}\right)^{2m}\right]
+ * \f}
+ *
+ * where \f$c_i\f$ are the zero-indexed modal coefficients, \f$N\f$ is the basis
+ * order, \f$\alpha\f$ determines much the coefficients are rescaled, and
+ * \f$m\f$ determines how aggressive/broad the filter is (lower values means
+ * filtering more coefficients). Setting \f$\alpha=36\f$ results in setting the
```
Comment (moxcodes) [inline src/NumericalAlgorithms/Spectral/Filtering.hpp:33]: again, perhaps a mention of 'zero' would be easier to understand.

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+                std::make_index_sequence<Spectral::maximum_number_of_points<
+                    Spectral::Basis::Chebyshev>>{});
+          default:
+            ERROR("Missing quadrature in exponential filter matrix action.");
+        }
+      default:
+        ERROR("Missing basis in exponential filter matrix action.");
+    };
+  }
+
```
Comment (moxcodes) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp:277]: There should probably be a quick code comment here explaining that this functions as a lazy static cache, that works via the function table technique, and reference other documentation if it exists (I don't recall whether you wrote anything more than that for the other one that you constructed).

File: tests/Unit/Evolution/DiscontinuousGalerkin/Test_Filtering.cpp [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    tuples::get<MockDistributedObjectsTag>(dist_objects)
+        .emplace(0, ActionTesting::MockDistributedObject<component>{
+                        db::create<typename component::simple_tags>(
+                            mesh, initial_scalar, initial_vector)});
+
+    MockRuntimeSystem runner{
+        {dg::Actions::ExponentialFilter<0, tmpl::list<Tags::ScalarVar>>{
+             alpha, half_power, false},
+         dg::Actions::ExponentialFilter<1, tmpl::list<Tags::VectorVar<Dim>>>{
+             2.0 * alpha, 2 * half_power, false}},
```
Comment (moxcodes) [inline tests/Unit/Evolution/DiscontinuousGalerkin/Test_Filtering.cpp]: test that disable for debug = true does nothing.

File: src/NumericalAlgorithms/Spectral/Filtering.hpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ * apply a stable exponential filter.
+ *
+ * The exponential filter rescales the coefficients according to:
+ *
+ * \f{align*}{
+ *  c_i\to c_i \exp\left[-\alpha \left(\frac{i}{N}\right)^{2m}\right]
+ * \f}
+ *
+ * where \f$c_i\f$ are the zero-indexed modal coefficients, \f$N\f$ is the basis
+ * order, \f$\alpha\f$ determines much the coefficients are rescaled, and
```
Comment (fmahebert) [inline src/NumericalAlgorithms/Spectral/Filtering.hpp]: much -> how much ?

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Filtering.cpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            local_approx(initial_modal_coeffs[i] *
+                         exp(-alpha * pow(i / basis_order, 2 * half_power))));
+    }
+  }
+}
+
+SPECTRE_TEST_CASE("Unit.Numerical.Spectral.ExponentialFilter",
+                  "[NumericalAlgorithms][Spectral][Unit]") {
+  std::vector<double> alphas{10.0, 20.0, 30.0, 40.0};
+  std::vector<unsigned> half_powers{2, 4, 8, 16};
```
Comment (fmahebert) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_Filtering.cpp]: can these two vectors be const?

File: src/NumericalAlgorithms/Spectral/Filtering.hpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace Spectral {
+/*!
+ * \ingroup SpectralGroup
+ * \brief Matrices for filtering spectral coefficients.
+ */
+namespace filtering {
+/*!
+ * \brief Returns a `Matrix` by which to multiply the modal coefficients to
+ * apply a stable exponential filter.
```
Comment (fmahebert) [inline src/NumericalAlgorithms/Spectral/Filtering.hpp:22]: The .cpp and the test both suggest this is returning a matrix to multiply the *nodal* coeffs.

I suggest changing modal -> nodal on line 21 (because in the brief I just want to know how to use the matrix), and also changing "coefficients" -> "modal coefficients" on line 24 below (because the more detailed discussion below should precisely define the filter in terms of its action on modal coeffs).

**Review by fmahebert — CHANGES_REQUESTED** [review]
Starting by reviewing one of the two commits; more later.

File: tests/Unit/NumericalAlgorithms/Spectral/Test_LegendreGauss.cpp [hunk 10]
```diff
@@ -128,32 +128,30 @@ SPECTRE_TEST_CASE("Unit.Numerical.Spectral.LegendreGauss.DiffMatrix",
...
 
 }  // namespace
 
-SPECTRE_TEST_CASE("Unit.Numerical.Spectral.LegendreGauss.SpectralToGridPoints",
+SPECTRE_TEST_CASE("Unit.Numerical.Spectral.LegendreGauss.ModalToNodal",
                   "[NumericalAlgorithms][Spectral][Unit]") {
-  SECTION("Check 1 point") {
-    test_spectral_to_grid_points_matrix(1, Matrix(1, 1, 1.));
-  }
+  SECTION("Check 1 point") { test_modal_to_nodal_matrix(1, Matrix(1, 1, 1.)); }
```
Comment (moxcodes) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_LegendreGauss.cpp]: This isn't really a change made in this PR but would it be better to scope and `INFO` here rather than `SECTION`?

**Review by moxcodes — COMMENTED** [review]
A few small comments, overall looks good to me.

File: tests/Unit/NumericalAlgorithms/Spectral/Test_LegendreGauss.cpp (hunk 10 above)
Comment (nilsdeppe) [reply tests/Unit/NumericalAlgorithms/Spectral/Test_LegendreGauss.cpp]: Yes, though that should probably be done as a revamp of all these low-level tests that were written long before we decided to not use `SECTION` :)

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: I don't understand what you mean by "these". Could you clarify? The switch statements are used to translate from the runtime mesh values compile-time template parameters. We need the mesh to be runtime so that it is easier to use and because we want the ability to have different meshes on different elements and element interfaces.

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Filtering.cpp (hunk 8 above)
Comment (nilsdeppe) [reply tests/Unit/NumericalAlgorithms/Spectral/Test_Filtering.cpp]: Why yes. I'll go into the box of shame now, thank you. :P

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+/// \cond
+template <size_t FilterIndex, typename TagsToFilterList>
+struct ExponentialFilter;
+/// \endcond
+
+/*!
+ * \ingroup DiscontinuousGalerkinGroup
+ * \brief Applies an exponential filter to the specified tags.
+ *
+ * Applies an exponential filter to each component of the tensors `FilterTags`.
```
Comment (fmahebert) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: FilterTags -> TagsToFilterList ?

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+ *
+ * \f{align*}{
+ *  c_i\to c_i \exp\left[-\alpha \left(\frac{i}{N}\right)^{2m}\right]
+ * \f}
+ *
+ * where \f$N\f$ is the basis order, \f$\alpha\f$ determines much the
+ * coefficients are rescaled, and \f$m\f$ (given by the `HalfPower` option)
+ * determines how aggressive/broad the filter is (lower values means filtering
+ * more coefficients). Setting \f$\alpha=36\f$ results in setting the highest
+ * coefficient to machine precision.
```
Comment (fmahebert) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: The above description of the exponential filter seems very similar from the filter matrix documentation. I have mixed feelings about it: one the one hand it's "code" duplication, but on the other hand it's useful to state it here. I'm just saying this so you think about the option of pointing to the documentation in the filter matrix. I have no strong feelings either way.

More concretely: while the filter matrix was inherently a 1D operation, here you could be filtering 2D or 3D data. If you want to keep the documentation, I would suggest clarifying that the equation you've given applies in each logical direction.

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+
+  using type = ExponentialFilter;
+  static std::string name() noexcept {
+    return "ExpFilter" + std::to_string(FilterIndex);
+  }
+
+  using options = tmpl::list<Alpha, HalfPower, DisableForDebugging>;
+  static constexpr OptionString help = {
+      "An exponential filter.\n"
+      // "The filter is applied in each direction to each tag."
```
Comment (fmahebert) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: Remove this line

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+      noexcept {
+    static const std::array<const Matrix& (*)(const Mesh<1>&, double, unsigned),
+                            sizeof...(Is)>
+        cache{{&filter_matrix_cache<BasisType, QuadratureType, Is>...}};
+    return gsl::at(cache, mesh.extents(0))(mesh, alpha_, half_power_);
+  }
+
+  double alpha_{36.0};
+  unsigned half_power_{16};
+  bool disable_for_debugging_{false};
```
Comment (fmahebert) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp:301]: ~~I don't believe you actually use this to disable the filtering :P~~

Edit: I see you addressed this just before I submitted my comments

**Comment by nilsdeppe** [comment]
Thanks for the feedback @fmahebert and @moxcodes. I pushed fixups. Please let me know if I missed anything or if you have any other suggestions :)

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+  }
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) noexcept;
+
+  bool operator==(const ExponentialFilter& rhs) const noexcept;
+
+ private:
+  template <size_t Dim>
+  const Matrix& filter_matrix(const Mesh<Dim>& mesh, size_t d) const noexcept {
```
Comment (fmahebert) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: I think you can eliminate the template parameter Dim in this function by taking as argument a `Mesh<1>`, and calling the function with `filter_matrix(mesh.slice_through(d))`. I'm not sure if this is worth doing.

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp (hunk 11 above)
Comment (nilsdeppe) [reply src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: Since I'm documenting the version that doesn't have the `List` at the end I'll omit that, but agreed :)

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp (hunk 12 above)
Comment (nilsdeppe) [reply src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: I did think about this and decided on wanting to have this be more user friendly and self-contained at the cost of some duplication. I'll discuss 2d and 3d, that's a very good point :)

**Comment by nilsdeppe** [comment]
@fmahebert I've pushed a fixup to address your comments. Let me know if you want anything else changed :)

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp (hunk 2 above)
Comment (moxcodes) [reply src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: I see - I have misunderstood the mesh interface, and thought `Mesh`es were also templated on basis and quadrature.

**Review by moxcodes — DISMISSED** [review]

**Comment by fmahebert** [comment]
Jenkins: run tests

**Comment by fmahebert** [comment]
LGTM

**Review by fmahebert — DISMISSED** [review]

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+   *
+   * \f{align*}{
+   *  c_i\to c_i \exp\left[-\alpha \left(\frac{i}{N}\right)^{2m}\right]
+   * \f}
+   */
+  struct HalfPower {
+    using type = unsigned;
+    static constexpr OptionString help =
+        "Half of the exponent in the generalized Gaussian";
+  };
```
Comment (hrueter) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp:134]: Why are you using half power as an option. I know that it was done this way in SpEC, but I think it is unnecessary and confusing.

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+        [&filter, &temp ](
+            const gsl::not_null<db::item_type<TagsToFilter>*>... filter_tensors,
+            const auto& local_mesh) noexcept {
+          const auto helper =
+              [&local_mesh, &filter, &temp ](const auto tensor) noexcept {
+            for (auto& component : *tensor) {
+              temp = 0.0;
+              apply_matrices(make_not_null(&temp), filter, component,
+                             local_mesh.extents());
+              component = temp;
```
Comment (hrueter) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: why are you using temp instead of the return by value version of `apply_matrices`?

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+  ExponentialFilter() = default;
+
+  ExponentialFilter(double alpha, unsigned half_power,
+                    bool disable_for_debugging) noexcept;
+
+  // Action part of the class
+  template <typename DbTags, typename... InboxTags, typename ArrayIndex,
+            typename ActionList, typename ParallelComponent,
+            typename Metavariables>
+  static std::tuple<db::DataBox<DbTags>&&> apply(
```
Comment (hrueter) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp:163]: I think in practice it might also be nice to have a filter action acting on the whole state vector. This would make filtering also more efficient, since we can filter the whole `Variables` in one go, instead of the individual components.

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 19]
```diff
@@ -0,0 +1,281 @@
+// Distributed under the MIT License.
```
Comment (hrueter) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp:1]: Minor nitpick:
Why do you use the gerund "Filtering" instead of just "Filter"? I think everywhere else we use the infinitive version in the file names.

File: src/NumericalAlgorithms/Spectral/Filtering.cpp [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+namespace filtering {
+Matrix exponential_filter(const Mesh<1>& mesh, const double alpha,
+                          const unsigned half_power) noexcept {
+  const Matrix& nodal_to_modal = Spectral::nodal_to_modal_matrix(mesh);
+  const Matrix& modal_to_nodal = Spectral::modal_to_nodal_matrix(mesh);
+  Matrix filter_matrix(mesh.number_of_grid_points(),
+                       mesh.number_of_grid_points(), 0.0);
+  const double order = mesh.number_of_grid_points() - 1.0;
+  for (size_t i = 0; i < mesh.number_of_grid_points(); ++i) {
+    filter_matrix(i, i) = exp(-alpha * pow(i / order, 2 * half_power));
```
Comment (hrueter) [inline src/NumericalAlgorithms/Spectral/Filtering.cpp:26]: This will be problematic if the number of grid points is 1 (i.e. order == 0), which is an allowed number for Gauss quadrature.
For this case I suggest `filter_matrix = 1.0`, which could be returned immediately.

File: tests/Unit/Evolution/DiscontinuousGalerkin/Test_Filtering.cpp [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+struct Metavariables {
+  using system = System<Dim>;
+  static constexpr bool local_time_stepping = true;
+  using component_list = tmpl::list<Component<Metavariables>>;
+  using const_global_cache_tag_list = tmpl::list<>;
+};
+
+template <size_t Dim, Spectral::Basis BasisType,
+          Spectral::Quadrature QuadratureType>
+void test_exponential_filter_action(const size_t min_pts, const double alpha,
```
Comment (hrueter) [inline tests/Unit/Evolution/DiscontinuousGalerkin/Test_Filtering.cpp]: Test for all number of points that are allowed by the quadrature.

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+class ExponentialFilter<FilterIndex, tmpl::list<TagsToFilter...>> {
+ public:
+  using const_global_cache_tags = tmpl::list<ExponentialFilter>;
+
+  /// \brief The value of `exp(-alpha)` is what the highest modal coefficient is
+  /// rescaled by.
+  struct Alpha {
+    using type = double;
+    static constexpr OptionString help =
+        "exp(-alpha) is rescaling of highest coefficient";
```
Comment (hrueter) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp:119]: Alpha should be positive
```suggestion
        "exp(-alpha) is rescaling of highest coefficient";
        static type lower_bound() noexcept { return 0.0; }
```

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+   * \brief Half of the exponent in the exponential.
+   *
+   * \f{align*}{
+   *  c_i\to c_i \exp\left[-\alpha \left(\frac{i}{N}\right)^{2m}\right]
+   * \f}
+   */
+  struct HalfPower {
+    using type = unsigned;
+    static constexpr OptionString help =
+        "Half of the exponent in the generalized Gaussian";
```
Comment (hrueter) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp:132]: I think zero should be excluded.
```suggestion
        "Half of the exponent in the generalized Gaussian";
        static type lower_bound() noexcept { return 1; }
```

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 24]
```diff
@@ -0,0 +1,281 @@
...
+ *
+ * \snippet DiscontinuousGalerkin/Test_Filtering.cpp action_list_example
+ *
+ * #### Action properties
+ *
+ * Uses:
+ * - ConstGlobalCache:
+ *   - ExponentialFilter
+ * - DataBox:
+ *   - Tags::Mesh
```
Comment (hrueter) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: Add dependence on volume_dim
```suggestion
 *   - Tags::Mesh
 * - System:
 *   - volume_dim
```

**Review by hrueter — COMMENTED** [review]
Can you add the action to the  ScalarWave, so that we have an example usage?

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 25]
```diff
@@ -0,0 +1,281 @@
...
+namespace dg {
+namespace Actions {
+/// \cond
+template <size_t FilterIndex, typename TagsToFilterList>
+struct ExponentialFilter;
+/// \endcond
+
+/*!
+ * \ingroup DiscontinuousGalerkinGroup
+ * \brief Applies an exponential filter to the specified tags.
```
Comment (hrueter) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp:58]: Add a reference to e.g. Hesthaven.

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp (hunk 16 above)
Comment (nilsdeppe) [reply src/Evolution/DiscontinuousGalerkin/Filtering.hpp:134]: I'm not sure what you're suggesting be changed. Things I can think of are:
1) The name of the option
2) That it be the power instead of half power
3) Not allowing people to set this option
4) Something else?

My responses to these are:
1) I'm open to suggestions, but I think the current one is very clear.
2) Using the power is wrong because the power needs to be even. Setting the half power prevents a user from ever getting it to be odd without having to check anything and is consistent with Hesthaven.
3) Since this controls how many modes are filtered I think it's a necessary option.
4) I'd need clarification to respond.

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp (hunk 17 above)
Comment (nilsdeppe) [reply src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: Because I don't want to reallocate memory for each tensor component.

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp (hunk 19 above)
Comment (nilsdeppe) [reply src/Evolution/DiscontinuousGalerkin/Filtering.hpp:1]: Because colloquially we refer to it as `Filtering` and `partial derivatives`, which is how I've interpreted the names being chosen.

**Comment by nilsdeppe** [comment]
@hrueter I think I've addressed all your comments that I can address without getting clarification.

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 26]
```diff
@@ -137,37 +166,55 @@ class ExponentialFilter<FilterIndex, tmpl::list<TagsToFilter...>> {
...
+                  vars,
+              const auto& local_mesh) noexcept {
+            *vars = apply_matrices(filter, *vars, local_mesh.extents());
+          },
+          mesh);
+    } else {
+      db::mutate<TagsToFilter...>(
+          make_not_null(&box),
+          [&filter](const gsl::not_null<
+                        db::item_type<TagsToFilter>*>... filter_tensors,
```
Comment (hrueter) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: like `TagsToFilter` using `tensors_to_filter` would be more clear.

File: tests/Unit/Evolution/DiscontinuousGalerkin/Test_Filtering.cpp [hunk 27]
```diff
@@ -0,0 +1,260 @@
...
+  CAPTURE(QuadratureType);
+  CAPTURE(disable_for_debugging);
+
+  using metavariables = Metavariables<Dim, FilterIndividually>;
+  using component = Component<metavariables>;
+  using MockRuntimeSystem = ActionTesting::MockRuntimeSystem<metavariables>;
+  using MockDistributedObjectsTag =
+      typename MockRuntimeSystem::template MockDistributedObjectsTag<component>;
+
+  for (size_t num_pts = min_pts;
```
Comment (hrueter) [inline tests/Unit/Evolution/DiscontinuousGalerkin/Test_Filtering.cpp]: Instead of passing `min_pts` just use `Spectral::minimum_number_of_points<BasisType,QuadratureType>`

File: tests/Unit/Evolution/DiscontinuousGalerkin/Test_Filtering.cpp [hunk 28]
```diff
@@ -191,15 +238,21 @@ void test_exponential_filter_creation() noexcept {
 SPECTRE_TEST_CASE("Unit.Evolution.dG.ExponentialFilter", "[Unit][Evolution]") {
   // Can't do a loop over different alpha and half_power values because matrices
   // are cached in the action.
-  const double alpha = 20.0;
+  const double alpha = 10.0;
```
Comment (hrueter) [inline tests/Unit/Evolution/DiscontinuousGalerkin/Test_Filtering.cpp:242]: Why did you change this? 20 was nice.

File: tests/InputFiles/ScalarWave/Input1DPeriodic.yaml [hunk 29]
```diff
@@ -39,6 +39,10 @@ StepChoosers:
   - Cfl:
       SafetyFactor: 0.2
 
+ExpFilter0:
+  Alpha: 12
```
Comment (hrueter) [inline tests/InputFiles/ScalarWave/Input1DPeriodic.yaml]: Why are you using `12` instead of the `36` proposed by Hesthaven?

**Review by hrueter — COMMENTED** [review]
I went through your new commits. You can squash as far as I am concerned.

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp (hunk 16 above)
Comment (hrueter) [reply src/Evolution/DiscontinuousGalerkin/Filtering.hpp:134]: I was concerned about point 2. Thanks.

**Comment by fmahebert** [comment]
I don't really like adding the filtering example in the scalar wave executable. While I agree it is useful to show how the filtering is compiled and specified in input files, I think it is misleading to add filtering to a problem that doesn't need it. I would like for the default executables and input files to provide reasonable starting points for numerical experiments, but now the first step to running experiments on the scalar wave will be to remove the filter...

I propose the following compromise:
- Add notes to the ScalarWave executable and input files, everywhere the filter appears, to clarify that filtering is not needed and should be removed in any numerical experiments.
- Open an issue to remove filtering from the ScalarWave executable. This task can be done after we have added a new example executable that solves a problem where filtering is actually necessary. It seems likely that we will eventually have such an executable.

Or, to show the filter in a pedagogically relevant way using current code, a more drastic change:
- Remove filtering from the ScalarWave examples.
- Add a new Burgers executable that uses (aggressive) filtering instead of the limiter to stabilize a shock.

Or perhaps alternative solutions?

File: src/NumericalAlgorithms/Spectral/Filtering.hpp [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+template <size_t>
+class Mesh;
+/// \endcond
+
+namespace Spectral {
+/*!
+ * \ingroup SpectralGroup
+ * \brief Matrices for filtering spectral coefficients.
+ */
+namespace filtering {
```
Comment (moxcodes) [inline src/NumericalAlgorithms/Spectral/Filtering.hpp:19]: What determines when we prefer a capitalized namespace vs a lower-case namespace? It seems to me we should choose one and try to be consistent.

**Comment by nilsdeppe** [comment]
@fmahebert @hrueter can you two please decide on what you want done here? You are currently both requesting opposite changes...

File: src/NumericalAlgorithms/Spectral/Filtering.hpp (hunk 30 above)
Comment (nilsdeppe) [reply src/NumericalAlgorithms/Spectral/Filtering.hpp:19]: About a year or so ago @kidder (and I think @wthrowe too) and I decided we are consistently inconsistent, don't have a preference for the casing, and don't care enough to change it.

**Comment by hrueter** [comment]
I approve @fmahebert's first suggestion.

File: tests/InputFiles/ScalarWave/Input1DPeriodic.yaml (hunk 29 above)
Comment (nilsdeppe) [reply tests/InputFiles/ScalarWave/Input1DPeriodic.yaml]: Hesthaven says to start with 36 and experiment. Specifically, you want to filter only as much as necessary, not more.

File: tests/Unit/Evolution/DiscontinuousGalerkin/Test_Filtering.cpp (hunk 28 above)
Comment (nilsdeppe) [reply tests/Unit/Evolution/DiscontinuousGalerkin/Test_Filtering.cpp:242]: 10 is easier for debugging because the values differ in a more significant digit than 20. It made it easier to figure out where bugs were in the code :)

**Comment by nilsdeppe** [comment]
@fmahebert @hrueter  @moxcodes I've squashed and addressed or replied to comments. Please let me know if you have any other requests :)

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Filtering.cpp [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+
+template <Spectral::Basis BasisType, Spectral::Quadrature QuadratureType>
+void test_exponential_filter(const size_t min_pts, const double alpha,
+                             const unsigned half_power,
+                             const double eps) noexcept {
+  Approx local_approx = Approx::custom().epsilon(eps).scale(1.0);
+  CAPTURE(BasisType);
+  CAPTURE(QuadratureType);
+  CAPTURE(eps);
+  for (size_t num_pts = min_pts;
```
Comment (hrueter) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_Filtering.cpp]: Here as well.
Instead of passing `min_pts` just use `Spectral::minimum_number_of_points<BasisType,QuadratureType>`

**Review by hrueter — CHANGES_REQUESTED** [review]

File: tests/Unit/NumericalAlgorithms/Spectral/Test_Filtering.cpp (hunk 31 above)
Comment (nilsdeppe) [reply tests/Unit/NumericalAlgorithms/Spectral/Test_Filtering.cpp]: :+1: done!

**Review by hrueter — DISMISSED** [review]

**Review by moxcodes — DISMISSED** [review]
Looks good.

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
@fmahebert please look at

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 32]
```diff
@@ -0,0 +1,328 @@
...
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      const Parallel::ConstGlobalCache<Metavariables>& cache,
+      const ArrayIndex& /*array_index*/, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    constexpr size_t volume_dim = Metavariables::system::volume_dim;
+    using evolved_vars_tag = typename Metavariables::system::variables_tag;
+    using evolved_vars = typename evolved_vars_tag::tags_list;
+    const ExponentialFilter& filter_helper =
+        Parallel::get<ExponentialFilter>(cache);
+    if (filter_helper.disable_for_debugging_) {
```
Comment (fmahebert) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: UNLIKELY

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 33]
```diff
@@ -0,0 +1,328 @@
...
+    }
+    const Mesh<volume_dim> mesh = db::get<::Tags::Mesh<volume_dim>>(box);
+    const Matrix empty{};
+    auto filter = make_array<volume_dim>(std::cref(empty));
+    for (size_t d = 0; d < volume_dim; d++) {
+      gsl::at(filter, d) =
+          std::cref(filter_helper.filter_matrix(mesh.slice_through(d)));
+    }
+
+    if (ExponentialFilter_detail::FilterAllEvolvedVars<
```
Comment (fmahebert) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp:188]: Do I correctly understand that this if-else block serves to optimize the case where the TagsToFilter are the evolved variables, by using the overload of apply_matrices that handles the entires vars at once? If so, I suggest adding a short comment to help the reader understand this is optimization. Perhaps: "In the case that the TagsToFilter are the evolved variables, we can optimize the code by applying the filter matrix to the entire evolved variables at once. In the general case, the matrix is applied to one tensor component at a time."

File: src/Evolution/DiscontinuousGalerkin/Filtering.hpp [hunk 34]
```diff
@@ -0,0 +1,328 @@
...
+            typename Metavariables>
+  static std::tuple<db::DataBox<DbTags>&&> apply(
+      db::DataBox<DbTags>& box,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      const Parallel::ConstGlobalCache<Metavariables>& cache,
+      const ArrayIndex& /*array_index*/, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    constexpr size_t volume_dim = Metavariables::system::volume_dim;
+    using evolved_vars_tag = typename Metavariables::system::variables_tag;
+    using evolved_vars = typename evolved_vars_tag::tags_list;
```
Comment (fmahebert) [inline src/Evolution/DiscontinuousGalerkin/Filtering.hpp]: I think this could be easier to parse if named `evolved_vars_tags_list` or `evolved_tags_list`, or something that makes it clearer this is a tmpl::list of tensor tags, rather than a variable type.

**Review by fmahebert — DISMISSED** [review]
I have a couple of very small suggestions that I think would help improve the readability of the code. Since this appears to be otherwise ready to merge, I am happy for Nils to address my comments in a follow-up PR if he likes.

**Review by fmahebert — APPROVED** [review]
LGTM 👍

**Review by kidder — APPROVED** [review]
